## Урок 1. ДЗ 
## Теория вероятности
### Случайные события. Условная вероятность. Формула Байеса.  Независимые испытания

In [113]:
import numpy as np
from math import factorial

**функция для подсчёта ПЕРЕСТАНОВОК из  𝑛  элементов
при перестановках важен порядок, но отличие от размещений в том, что применяются все имеющиеся  𝑛  элементы:**

$$P_n = n!$$

In [114]:
def permutations(n):
    return int(factorial(n))

**функция для подсчёта РАЗМЕЩЕНИЙ k-элементов  из n-элементов .
При размещениях порядок важен, поэтому вариантов размещения может быть больше, чем сочетаний при заданных  𝑘  и  𝑛 .**
$$A_n^k = \frac{n!}{(n - k)!}$$

In [116]:
def arrangements(n, k):
    return int(factorial(n) / factorial(n - k))

**функция для подсчёта СОЧЕТАНИЙ k-элементов выбранных из n-элементов (в сочетаниях порядок не важен):**

$$C_n^k = \frac{n!}{k!(n - k)!}$$

In [115]:
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

**ЗАДАЧА 1.**\
Из колоды в 52 карты извлекаются случайным образом 4 карты.\
a) Найти вероятность того, что все карты – крести.\
б) Найти вероятность, что среди 4-х карт окажется хотя бы один туз.

In [134]:
# Вначале воспользуемся формулой для вычисления количества всех сочетаний 4-х карт из 52:

n = (factorial(52) / (factorial(4) * factorial(52 - 4)))
n

270725.0

In [118]:
# крестей как и других мастей в колоде по: 52/4 = 13 карт.
# найдём сочетания 4 карт ко всем сочетаниям карт одной масти, к примеру крестей.
m = (factorial(13) / (factorial(4) * factorial(13 - 4)))
m

715.0

In [119]:
# Р(А) = m/n

print(f'Вероятность того, что все карты - крести = {m/n :.4f}')

Вероятность того, что все карты - крести = 0.0026


In [133]:
# Тузов в колоде всего 4. Вероятность выбрать из колоды хотя бы один туз:
n2 = (factorial(52) / (factorial(1) * factorial(52 - 1)))
n2

52.0

In [139]:
# Р(А) = m/n
print(f'Вероятность того, что в выбранных крестях окажется один туз = {0.0026/n2 :.5f}')

Вероятность того, что в выбранных крестях окажется один туз = 0.00005


**ЗАДАЧА 2.**\
На входной двери подъезда установлен кодовый замок, содержащий десять кнопок с цифрами от 0 до 9.\
Код содержит три цифры, которые нужно нажать одновременно.\
Какова вероятность того, что человек, не знающий код, откроет дверь с первой попытки?

**решение:**  число всех возможных сочетаний 3-х обьектов из 10 находится по формуле сочетания:

$$C_n^k = \frac{n!}{k!(n - k)!}$$

In [120]:
(factorial(10) / (factorial(3) * factorial(10 - 3)))

120.0

In [121]:
# либо используя функцию
combinations(10, 3)

120

**ответ:**  вероятность угадать с первого раза: 1 из 120

**ЗАДАЧА 3.**\
В ящике имеется 15 деталей, из которых 9 окрашены.\
Рабочий случайным образом извлекает 3 детали.\
Какова вероятность того, что все извлеченные детали окрашены?

In [122]:
#  Найдём все возможные сочетания 3 деталей ко всем имеющимся деталям. Это все равновероятные исходы.

m = (factorial(15) / (factorial(3) * factorial(15 - 3)))
m

455.0

In [123]:
#  Найдём все возможное сочетание 3-х деталей из 9-и окрашенных деталей.

n = (factorial(9) / (factorial(3) * factorial(9 - 3)))
n

84.0

In [124]:
# определим вероятность выбора только окрашенных деталей из всех возможных.
P = n/m
P

0.18461538461538463

In [125]:
# либо используя функции:

P = combinations(9, 3)/combinations(15, 3)
print(f'Вероятность выбрать 3 окрашенные детали: {P :.2f}')

Вероятность выбрать 3 окрашенные детали: 0.18


**ЗАДАЧА 4.**\
В лотерее 100 билетов. Из них 2 выигрышных.\
Какова вероятность того, что 2 приобретенных билета окажутся выигрышными?

**решение:** имеем различные сочетания независящих друг от друга 2-х различных предметов из 100. Применяем формулу сочетания.

$$C_n^k = \frac{n!}{k!(n - k)!}$$

In [126]:
С = (factorial(100) / (factorial(2) * factorial(100 - 2)))
С

4950.0

In [127]:
# либо используя функцию:
combinations(100, 2)

4950

**ответ:** вероятность угадать 2 выигрышных билета из 100:  1 из 4950

**ЗАДАЧА 5.**\
На соревновании по биатлону один из трех спортсменов стреляет и попадает в мишень.\
Вероятность попадания для первого спортсмена равна 0.9, для второго — 0.8, для третьего — 0.6.\
Найти вероятность того, что выстрел произведен:\
a). первым спортсменом\
б). вторым спортсменом\
в). третьим спортсменом.

**решение**

пусть:

Событие А1 - Первый стрелок попал в цель.\
Событие А2 - Второй стрелок попал в цель.\
Событие А3 - Третий стрелок попал в цель.

 известно:
 
вероятность попадания для первого  спортсмена P(A1) = p1 = 0.9\
вероятность попадания для второго  спортсмена P(A2) = p2 = 0.8\
вероятность попадания для третьего спортсмена P(A3) = p3 = 0.6

по условию кто то из них попал, поэтому общая вероятность = 100%\
P(A1)+P(A2)+P(A3) = 1 

подсчитаем вероятности НЕпопадания для каждого спортсмена:\
q1 = 1 - 0.9 = 0.1\
q2 = 1 - 0.8 = 0.2\
q3 = 1 - 0.6 = 0.4

теперь рассмотрим все варианты исхода выстрела


$$P1=\
P(A1)⋅P(\overline{A2})⋅P(\overline{A3})+P(\overline{A1})⋅P(A2)⋅P(\overline{A3})+P(\overline{A1})⋅P(\overline{A2})⋅P(A3)$$

P1=p1⋅q2⋅q3+q1⋅p2⋅q3+q1⋅q2⋅p3 == 0,9 ⋅ 0,2 ⋅ 0,4 + 0,1 ⋅ 0,8 ⋅ 0,4 + 0,1 ⋅ 0,2 ⋅ 0,6


In [128]:
# непонятное решение )))  моё )))
#P1 = 0.9 * 0.2 * 0.4 + 0.1 * 0.8 * 0.4 + 0.1 * 0.2 * 0.6
#P1

0.116

In [1]:
# Общая вероятность
A = 1/3 * 0.9 + 1/3 * 0.8 + 1/3 * 0.6
A

0.7666666666666666

In [2]:
# для первого спортсмена

(1/3 * 0.9) / A

0.391304347826087

In [3]:
# для второго спортсмена

(1/3 * 0.8) / A

0.3478260869565218

In [4]:
# для третьего спортсмена

(1/3 * 0.8) / A

0.3478260869565218